lib/shopify_app/controller_concerns/login_protection.rb

```diff
@@ -27,12 +27,30 @@ def shopify_session
     end
 
     def shop_session
-      return unless session[:shopify]
-      @shop_session ||= ShopifyApp::SessionRepository.retrieve(session[:shopify])
+      if ShopifyApp.configuration.per_user_tokens?
+        return unless session[:shopify_user]
+        @shop_session ||= ShopifyApp::SessionRepository.retrieve(session[:shopify_user]['id'])
+      else
+        return unless session[:shopify]
+        @shop_session ||= ShopifyApp::SessionRepository.retrieve(session[:shopify])
+      end
     end
 
-    def login_again_if_different_shop
+    def login_again_if_different_user_or_shop
+      if ShopifyApp.configuration.per_user_tokens?
+        invalid_session_data = session[:user_session].nil? || params[:session].nil? # session data was not sent/stored correctly
+        sessions_do_not_match = session[:user_session] != params[:session] # current user is different from stored user
+
+        if invalid_session_data || sessions_do_not_match
+          clear_session = true   
+        end
+      end
+
       if shop_session && params[:shop] && params[:shop].is_a?(String) && (shop_session.domain != params[:shop])
```

Comment:
Could this be in an `else`, as it only applies to the shop based token case?

Reply:
Line 49?

Reply:
It's minor, but:
```suggestion
      else
        if shop_session && params[:shop] && params[:shop].is_a?(String) && (shop_session.domain != params[:shop])
```

Reply:
Oh - line 49 needs to apply regardless of whether you are per-user or per-shop. What this does is do an additional check for the user session if you are in per-user mode.